### PR TITLE
Improve model cutting

### PR DESCRIPTION
**Pull request by mvafin**
### Details:
 - *If part of model is cut, conversion rules must not run on it*

### Tickets:
 - *None* - general improvement

### CODE REVIEW

File: ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/place.hpp [hunk 1]
```diff
@@ -170,7 +170,7 @@ namespace ngraph
             Ptr get_target_tensor(const std::string& outputName) const override;
             Ptr get_target_tensor(const std::string& outputName,
                                   int outputPortIndex) const override;
-
+            int state = 0; // used to determine cut nodes
```
Comment (ilyachur) [inline ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/place.hpp:173]: Why did you add a class member to public section?

File: ngraph/frontend/paddlepaddle/src/model.cpp [hunk 2]
```diff
@@ -226,8 +233,114 @@ namespace ngraph
...
+            std::queue<OpPlacePDPD*> q;
+            for (auto& input : getInputs())
+            {
+                pdpd::add_outputs_to_queue(q, input);
+            }
+            while (!q.empty())
+            {
+                auto p_op = q.front();
+                q.pop();
+                for (auto map_pair : p_op->get_output_ports())
```
Comment (ilyachur) [inline ngraph/frontend/paddlepaddle/src/model.cpp:317]: Looks like in cycles you can use `const auto&` or `auto&`.

File: ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/place.hpp (hunk 1 above)
Comment (mvafin) [reply ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/place.hpp]: removed

File: ngraph/frontend/paddlepaddle/src/model.cpp [hunk 3]
```diff
@@ -368,6 +446,7 @@ namespace ngraph
         void InputModelPDPD::InputModelPDPDImpl::overrideAllInputs(
             const std::vector<Place::Ptr>& inputs)
         {
+            m_graph_changed = true;
```
Comment (nosovmik) [inline ngraph/frontend/paddlepaddle/src/model.cpp:449]: why `determine_cut_nodes` is not called here, but delayed?

File: ngraph/frontend/paddlepaddle/src/frontend.cpp [hunk 4]
```diff
@@ -127,7 +119,7 @@ namespace ngraph
                 parameter_nodes.push_back(param);
             }
 
-            const auto& op_places = model->getOpPlaces();
+            const auto& op_places = model->returnOpPlaces();
```
Comment (nosovmik) [inline ngraph/frontend/paddlepaddle/src/frontend.cpp:122]: Can we probably do `determine_cut_nodes` here and use 'const' methods after this deferred update?

File: ngraph/frontend/paddlepaddle/src/model.cpp [hunk 5]
```diff
@@ -228,6 +235,77 @@ namespace ngraph
...
+            if (m_graph_changed)
+            {
+                determine_cut_nodes();
+                m_graph_changed = false;
+            }
+            return m_op_places;
+        }
+
+        void InputModelPDPD::InputModelPDPDImpl::determine_cut_nodes()
+        {
```
Comment (nosovmik) [inline ngraph/frontend/paddlepaddle/src/model.cpp:250]: Hard to review without tests. Did you test case when user cuts some input which is important to produce some output?

File: ngraph/frontend/paddlepaddle/src/model.cpp [hunk 6]
```diff
@@ -228,6 +235,77 @@ namespace ngraph
...
+                                        tensor->get_producing_operation());
+                                if (pdpd_op)
+                                {
+                                    visited.insert(pdpd_op.get());
+                                    q.push(pdpd_op.get());
+                                }
+                            }
+                            catch (GeneralFailure)
+                            {
+                                // tensor might not have producer op
```
Comment (nosovmik) [inline ngraph/frontend/paddlepaddle/src/model.cpp:290]: should we throw an error here if it is not tensor/constant?

File: ngraph/frontend/paddlepaddle/src/model.cpp (hunk 3 above)
Comment (mvafin) [reply ngraph/frontend/paddlepaddle/src/model.cpp:449]: No need to do it here. It is only needed when conversion started.

File: ngraph/frontend/paddlepaddle/src/frontend.cpp (hunk 4 above)
Comment (mvafin) [reply ngraph/frontend/paddlepaddle/src/frontend.cpp:122]: We can, but it makes more sense to have it done internally.

File: ngraph/frontend/paddlepaddle/src/model.cpp (hunk 5 above)
Comment (mvafin) [reply ngraph/frontend/paddlepaddle/src/model.cpp:250]: I will add such test

File: ngraph/frontend/paddlepaddle/src/model.cpp (hunk 6 above)
Comment (mvafin) [reply ngraph/frontend/paddlepaddle/src/model.cpp:290]: No, it is not a problem if some tensor doesn't have a producer.

File: ngraph/frontend/paddlepaddle/src/model.cpp [hunk 7]
```diff
@@ -228,6 +235,77 @@ namespace ngraph
...
+                    q.push(pdpd_output_op.get());
+                }
+            }
+            while (!q.empty())
+            {
+                auto p_op = q.front();
+                q.pop();
+                for (auto map_pair : p_op->get_input_ports())
+                {
+                    for (auto port : map_pair.second)
```
Comment (itikhono) [inline ngraph/frontend/paddlepaddle/src/model.cpp:271]: const ref?

File: ngraph/frontend/paddlepaddle/src/model.cpp [hunk 8]
```diff
@@ -228,6 +235,77 @@ namespace ngraph
...
+                        std::dynamic_pointer_cast<OpPlacePDPD>(output->get_producing_operation());
+                    visited.insert(pdpd_output_op.get());
+                    q.push(pdpd_output_op.get());
+                }
+            }
+            while (!q.empty())
+            {
+                auto p_op = q.front();
+                q.pop();
+                for (auto map_pair : p_op->get_input_ports())
```
Comment (itikhono) [inline ngraph/frontend/paddlepaddle/src/model.cpp:269]: const ref?

File: ngraph/frontend/paddlepaddle/src/model.cpp (hunk 6 above)
Comment (itikhono) [reply ngraph/frontend/paddlepaddle/src/model.cpp:290]: Do you have an example (tensor without producer)?
Looks odd

File: ngraph/frontend/paddlepaddle/src/model.cpp [hunk 9]
```diff
@@ -228,6 +235,77 @@ namespace ngraph
...
+            {
+                determine_cut_nodes();
+                m_graph_changed = false;
+            }
+            return m_op_places;
+        }
+
+        void InputModelPDPD::InputModelPDPDImpl::determine_cut_nodes()
+        {
+            // Marking nodes from inputs to outputs
```
Comment (itikhono) [inline ngraph/frontend/paddlepaddle/src/model.cpp:251]: Do we still need this comment?

File: ngraph/frontend/paddlepaddle/src/model.cpp (hunk 6 above)
Comment (mvafin) [reply ngraph/frontend/paddlepaddle/src/model.cpp:290]: It is similar to outputs, there are operations with optional inputs, they may have tensor, but not operation producing it.

File: ngraph/frontend/paddlepaddle/src/model.cpp (hunk 9 above)
Comment (mvafin) [reply ngraph/frontend/paddlepaddle/src/model.cpp:251]: No, will delete

File: ngraph/frontend/paddlepaddle/src/model.cpp [hunk 10]
```diff
@@ -228,6 +235,77 @@ namespace ngraph
...
+                        {
+                            try
+                            {
+                                std::shared_ptr<OpPlacePDPD> pdpd_op =
+                                    std::dynamic_pointer_cast<OpPlacePDPD>(
+                                        tensor->get_producing_operation());
+                                if (pdpd_op)
+                                {
+                                    visited.insert(pdpd_op.get());
+                                    q.push(pdpd_op.get());
```
Comment (itikhono) [inline ngraph/frontend/paddlepaddle/src/model.cpp:285]: please check pdpd_op already in visited

File: ngraph/frontend/paddlepaddle/src/model.cpp [hunk 11]
```diff
@@ -228,6 +235,77 @@ namespace ngraph
...
+                            catch (GeneralFailure)
+                            {
+                                // tensor might not have producer op
+                                continue;
+                            }
+                        }
+                    }
+                }
+            }
+            std::vector<std::shared_ptr<OpPlacePDPD>> new_op_places;
```
Comment (itikhono) [inline ngraph/frontend/paddlepaddle/src/model.cpp:297]: I guess we can create a vector in at the begging of the func and fill it during traversal

**Review by itikhono — APPROVED** [review]

File: ngraph/frontend/paddlepaddle/src/model.cpp [hunk 12]
```diff
@@ -228,6 +235,67 @@ namespace ngraph
...
+                                visited.insert(pdpd_op.get());
+                                q.push(pdpd_op.get());
+                                new_op_places.push_back(pdpd_op);
+                            }
+                        }
+                    }
+                }
+            }
+            std::reverse(new_op_places.begin(), new_op_places.end());
+            m_op_places.swap(new_op_places);
```
Comment (itikhono) [inline ngraph/frontend/paddlepaddle/src/model.cpp:296]: I guess we also need to update m_var_places

File: ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/model.hpp [hunk 13]
```diff
@@ -19,7 +19,7 @@ namespace ngraph
             class InputModelPDPDImpl;
             std::shared_ptr<InputModelPDPDImpl> _impl;
 
-            std::vector<std::shared_ptr<OpPlacePDPD>> getOpPlaces() const;
+            std::vector<std::shared_ptr<OpPlacePDPD>> returnOpPlaces();
```
Comment (ilyachur) [inline ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/model.hpp:22]: Do you start to modify class in this method?

File: ngraph/frontend/paddlepaddle/src/model.cpp (hunk 12 above)
Comment (mvafin) [reply ngraph/frontend/paddlepaddle/src/model.cpp:296]: we agreed to not do that

File: ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/model.hpp (hunk 13 above)
Comment (mvafin) [reply ngraph/frontend/paddlepaddle/include/paddlepaddle_frontend/model.hpp:22]: yes, because if model is cut we need to return smaller set of places

File: ngraph/frontend/paddlepaddle/src/model.cpp [hunk 14]
```diff
@@ -228,6 +235,67 @@ namespace ngraph
...
+            std::vector<std::shared_ptr<OpPlacePDPD>> new_op_places;
+            new_op_places.reserve(m_op_places.size());
+            // Marking nodes from outputs to inputs/constants
+            for (const auto& output : getOutputs())
+            {
+                if (!output->is_input())
+                {
+                    auto pdpd_output_op =
+                        std::dynamic_pointer_cast<OpPlacePDPD>(output->get_producing_operation());
+                    if (pdpd_output_op && !visited.count(pdpd_output_op.get()))
```
Comment (nosovmik) [inline ngraph/frontend/paddlepaddle/src/model.cpp:262]: do we need this `!visited.count`? It is for case when getOutputs return duplicates?

Comment (mvafin) [reply ngraph/frontend/paddlepaddle/src/model.cpp:262]: Yes, for the rare or maybe improbable case when several outputs point to same op.

File: ngraph/frontend/paddlepaddle/src/model.cpp [hunk 15]
```diff
@@ -228,6 +235,67 @@ namespace ngraph
...
+            InputModelPDPD::InputModelPDPDImpl::determine_cut_nodes() const
+        {
+            std::queue<OpPlacePDPD*> q;
+            std::unordered_set<OpPlacePDPD*> visited;
+            std::vector<std::shared_ptr<OpPlacePDPD>> new_op_places;
+            new_op_places.reserve(m_op_places.size());
+            // Marking nodes from outputs to inputs/constants
+            for (const auto& output : getOutputs())
+            {
+                if (!output->is_input())
```
Comment (nosovmik) [inline ngraph/frontend/paddlepaddle/src/model.cpp:258]: Is this check really needed? If yes, should there be an 'assert' instead?

File: ngraph/frontend/paddlepaddle/src/model.cpp (hunk 14 above)
Comment (nosovmik) [reply ngraph/frontend/paddlepaddle/src/model.cpp:262]: Also is check `if (pdpd_output_op)` needed? Should there be an assert instead?

File: ngraph/frontend/paddlepaddle/src/model.cpp [hunk 16]
```diff
@@ -388,6 +458,7 @@ namespace ngraph
         void InputModelPDPD::InputModelPDPDImpl::extractSubgraph(
             const std::vector<Place::Ptr>& inputs, const std::vector<Place::Ptr>& outputs)
         {
+            m_graph_changed = true;
```
Comment (nosovmik) [inline ngraph/frontend/paddlepaddle/src/model.cpp:461]: may not be needed here as `overrideAllInputs/overrideAllOutputs` already set this internally

File: ngraph/frontend/paddlepaddle/src/model.cpp [hunk 17]
```diff
@@ -419,6 +490,7 @@ namespace ngraph
 
         void InputModelPDPD::InputModelPDPDImpl::setTensorValue(Place::Ptr place, const void* value)
         {
+            m_graph_changed = true;
```
Comment (nosovmik) [inline ngraph/frontend/paddlepaddle/src/model.cpp:493]: Minor: if I comment this out - existing tests will be passed. Do we need new test verifying need of this flag on this line?

File: ngraph/frontend/paddlepaddle/src/model.cpp [hunk 18]
```diff
@@ -378,6 +447,7 @@ namespace ngraph
         void InputModelPDPD::InputModelPDPDImpl::overrideAllOutputs(
             const std::vector<Place::Ptr>& outputs)
         {
+            m_graph_changed = true;
```
Comment (nosovmik) [inline ngraph/frontend/paddlepaddle/src/model.cpp:450]: Minor: if I comment this out - existing tests will be passed. Do we need new test verifying need of this flag on this line?

File: ngraph/frontend/paddlepaddle/src/model.cpp (hunk 15 above)
Comment (mvafin) [reply ngraph/frontend/paddlepaddle/src/model.cpp:258]: This check is needed to test if we have Parameter->Result graph, it would look strange, but don't think we should throw in that case.

File: ngraph/frontend/paddlepaddle/src/model.cpp (hunk 16 above)
Comment (mvafin) [reply ngraph/frontend/paddlepaddle/src/model.cpp:461]: True, but what if those methods get changed? I don't think it is a big problem to have this flag set here too.

File: ngraph/frontend/paddlepaddle/src/model.cpp (hunk 17 above)
Comment (mvafin) [reply ngraph/frontend/paddlepaddle/src/model.cpp:493]: New behavior shouldn't change existing tests, to really see the difference we need to add a partially unsupported model to tests.

File: ngraph/frontend/paddlepaddle/src/model.cpp (hunk 18 above)
Comment (mvafin) [reply ngraph/frontend/paddlepaddle/src/model.cpp:450]: New behavior shouldn't change existing tests, to really see the difference we need to add a partially unsupported model to tests.

**Review by nosovmik — APPROVED** [review]

**Review by ilyachur — APPROVED** [review]
